include/revng/TupleTree/DiffError.h

```diff
@@ -0,0 +1,162 @@
+
+  IDContainer Ids;
+
+public:
+  explicit RevngErrorBase(const IDContainer &Ids) :
+    Reasons(), Ids(std::move(Ids)) {}
+  RevngErrorBase() = default;
+
+  static llvm::Error concatenate(llvm::Error SourceError,
+                                 llvm::StringRef NewReason,
+                                 std::unique_ptr<LocationBase> Loc) {
+
+    if (not SourceError.isA<RevngErrorBase>())
+      return SourceError;
+
+    llvm::Error ToReturn = llvm::Error::success();
+    return llvm::handleErrors(std::move(SourceError),
+                              [&](std::unique_ptr<revng::RevngErrorBase>
+                                    RevngError) {
+                                RevngError->Reasons.push_back(NewReason.str());
+                                RevngError->Locations.push_back(std::move(Loc));
+                                return llvm::Error(std::move(RevngError));
+                              });
+  }
+
+  // Returns the class ID for this type.
+  static const void *classID() { return &ID; }
+
+  iterator begin() { return Reasons.begin(); }
+
+  iterator end() { return Reasons.end(); }
+
+  [[nodiscard]] const_iterator begin() const { return Reasons.begin(); }
+
+  [[nodiscard]] const_iterator end() const { return Reasons.end(); }
+
+  size_t size() const { return Reasons.size(); }
+
+  const std::string &getMessage(size_t I) const { return Reasons[I]; }
+  std::string getLocation(size_t I) const { return Locations[I]->toString(); }
+
+  virtual bool isA(const void *const ClassID) const {
+    return llvm::find(Ids, ClassID) != Ids.end();
+  }
+
+public:
+  std::error_code convertToErrorCode() const final {
+    return llvm::inconvertibleErrorCode();
+  }
+  void log(llvm::raw_ostream &OS) const final {
+    OS << "Diff Errors\n";
+    for (auto &Reason : Reasons) {
+      OS << "\t" << Reason << "\n";
+    }
+  }
+};
+
+template<typename Derived, typename Location>
+class RevngError : public RevngErrorBase {
+private:
+  static RevngErrorBase::IDContainer getClassIds() {
+    RevngErrorBase::IDContainer Ids({ llvm::ErrorInfoBase::classID(),
+                                      Derived::classID(),
+                                      RevngError::classID(),
+                                      RevngErrorBase::classID() });
+    return Ids;
+  }
+
+public:
+  using LocationType = Location;
+  inline static char ID = '0';
+  RevngError(const RevngErrorBase::IDContainer &Ids = getClassIds()) :
+    RevngErrorBase(Ids) {}
+  RevngError(llvm::StringRef Reason,
+             const Location &ReasonLocation,
+             const RevngErrorBase::IDContainer &Ids = getClassIds()) :
+    RevngErrorBase(Ids) {
+    addReason(Reason, ReasonLocation);
+  }
+
+  void addReason(llvm::StringRef NewReason, const Location &ReasonLocation) {
+    Reasons.push_back(NewReason.str());
+    Locations.push_back(std::make_unique<Location>(ReasonLocation));
+  }
+
+  void consumeAndAppend(llvm::Error Error, const Location &ReasonLocation) {
```

Comment:
I see no reason to eat other `llvm::Error`s, this is not a collection of `llvm::Error`s but one error with multiple reasons and locations.
If the caller wants to do this, it should be his responsibility (maybe with an helper function), but I don't see this as a feature of `RevngError`.

Reply:
i am fiddling with this one but there is a annoying thing. When you create a error you do `llvm::createError<ErrorType>`  which returns a llvm::Error not a ErrorType, thus it's very uncomfortable to accumulate errors, because  you can't cast a error without consuming it.

Reply:
I didn't express myself clearly: I don't want this method not to consume the error, I don't want this method at all `:D`

`RevngError` is not a bin of arbitrary errors, it's a single error multiple reasons and locations.
I'd like to keep this very simple and only let people add reasons and locations in a plain way, without involving any other `llvm::Error`.

Reply:
the issue is that we do need it something like this

if you have a situation like 

```cpp
llvm::Error maybeError = llvm::Error::success();

while (something())
{
  auto maybeError2 = somethingFailable();
  if (maybeError2)
    maybeError = compose(std::move(maybeError), std::move(maybeError2));

}

if (maybeError)
  return std::move(maybeError);
```

Reply:
You can just use a vector of reasons and create the error at the end.
Our error is not composable with other errors, it's just an error with multiple reasons.

We could force to be composable only with error with the same type, but it's just simpler to have a vector and construct the error at the end.

The way we will use this is to have some sort of diagnostics class where we accumulate problems, we won't be combining things as you're suggesting.